### ISSUE TITLE
Broken links when setting a sub-path in app.baseUrl configuration

### ISSUE THREAD

**Issue by blaryjp** [backstage/backstage#6591]
## Expected Behavior

If we set a `app.baseUrl` with a sub-path, all the links must use it and correctly redirect.

## Current Behavior

Backstage works well if we don't put a sub-path in the `app.baseUrl` config; otherwise, many links are broken because they don't have the sub-path.

For example (this is not the only one, many other links are also broken):
If I put `baseUrl: http://localhost:3000/john/doe/`, all the links in the Catalog table list are broken, they goes to `/catalog/...` instead of `/john/doe/catalog/...`. 
I dig in the code, and I think that I've identified the problem: [see here](https://github.com/backstage/backstage/blob/master/plugins/catalog-react/src/components/EntityRefLink/EntityRefLink.tsx#L65) : this is probably due to the `generatePath` that doesn't put the base path when generating the URL.

## Possible Solution

Encapsulate `generatePath` to append the base path? Or be able to set it globally?

## Steps to Reproduce

1. Put `app.baseUrl: http://localhost:3000/john/doe/` in your app-config.yaml.
2. Launch Backstage and go to the /john/doe/catalog page
3. You can see that the catalog items have wrong URL paths

## Your Environment

- NodeJS Version (v12): node v14
- Operating System and Version (e.g. Ubuntu 14.04): Ubuntu 18.04
- Browser Information: Chromium 91

**Comment by henry232323** [comment]
This is an extension of [this bug](https://github.com/ReactTraining/react-router/issues/7216) in React Router, which is currently pending a [pull request](https://github.com/ReactTraining/react-router/pull/7462). When using `Link`'s, they will resolve normally unless they are absolute paths, i.e. prefixed with `/`

**Comment by blaryjp** [comment]
Thank you @henry232323 ! I'll take a look

**Comment by blaryjp** [comment]
Hello @henry232323  : I checked the PR, but it seems that it'll not fix this problem.
The problem here is the `generatePath` function, and the code in react-router (in that PR) don't fix it: https://github.com/ReactTraining/react-router/blob/f3ae557471cd0402b4da10fc21a57bef57956dc1/packages/react-router/index.tsx#L755

What do you think?

**Comment by henry232323** [comment]
Hmm, I recall hearing something about the rewritten path generation when researching the issue, not sure where though.
Have you given the PR a try to see if it resolves the issue in any capacity? I'm currently working through seeing if I can test out that PR but am running into some blockers, I'll update here if I am successful.

**Comment by henry232323** [comment]
I had success building the PR branch locally, then applying it as a patch with [patch-package](https://www.npmjs.com/package/patch-package). After some testing, it appears to fix every link I can find except using the search bar redirects back to root, and the register existing component button gives a blank page that yields no errors. I had to implement a BrowserRouter in App.tsx
```ts
const app = createApp({
    apis,
    components: {
        Router: props => (
            <BrowserRouter basename="/my/path">{props.children}</BrowserRouter>
        ),
    },
    ...
});
```
I also had to add this to `index.ts` and `scaffolder.ts` respectively to update their discovery to accommodate for the base url.
```ts
const discovery = SingleHostDiscovery.fromConfig(config, { basePath: '/my/path/api' }); // index.ts in makeCreateEnv
const discovery = SingleHostDiscovery.fromConfig(config, { basePath: '/my/path' }); // scaffolder.ts in createPlugin
```

**Comment by blaryjp** [comment]
Hello @henry232323 , thank you very much for your investigation.
I tried what you did: Indeed, it fixes the links inside the catalog table :).
I applied the modifications that you've written, but I still have the errors that you said (search, register existing component,...).
I don't understand.

Also, I've some others errors: the tab "docs" in a service have an error (http://localhost:3000/my/path/catalog/default/component/artist-lookup/docs for example):

```
Error: Cannot route to routeRef{type=absolute,id=Docs} with parent routeRef{type=absolute,id=Entity} as it has parameters
    at eval (webpack-internal:///../../node_modules/@backstage/core-app-api/dist/index.esm.js:1905:13)
    at Array.map (<anonymous>)
    at resolveBasePath (webpack-internal:///../../node_modules/@backstage/core-app-api/dist/index.esm.js:1899:58)
    at RouteResolver.resolve (webpack-internal:///../../node_modules/@backstage/core-app-api/dist/index.esm.js:1923:22)
    at eval (webpack-internal:///../../node_modules/@backstage/core-plugin-api/dist/index.esm.js:371:102)
    at mountMemo (webpack-internal:///../../node_modules/@hot-loader/react-dom/cjs/react-dom.development.js:15443:19)
    at Object.useMemo (webpack-internal:///../../node_modules/@hot-loader/react-dom/cjs/react-dom.development.js:15739:16)
    at useMemo (webpack-internal:///../../node_modules/react/cjs/react.development.js:1521:21)
    at wrappedHook (webpack-internal:///../../node_modules/react-hot-loader/dist/react-hot-loader.development.js:2798:14)
    at useRouteRef (webpack-internal:///../../node_modules/@backstage/core-plugin-api/dist/index.esm.js:371:74)
```
Also for some plugins (github pull requests, sentry, ...).

And finaly, when I refresh any (sub) page, I've got a 404.

Do you also have these errors?

**Comment by henry232323** [comment]
I'm encountering the same 404 refresh issue when I host locally, but not in my deployed environment for some reason. The register component issue seems to be almost the opposite issue, redirecting to `/create/catalog-import` instead of `/catalog-import`. A lot of this seems to stem from overall lacking support for these extended base urls, which may take some proper changes within Backstage itself. The code I provided above was all given to me in the Backstage support Discord server, which I think is worth joining for discussing some of these issues in more depth

**Comment by blaryjp** [comment]
Hey @henry232323 , thank you again for your investigations.
You're right, it'll be easier on Discord.

**Comment by timbonicus** [comment]
Fixed by #6749.

**Comment by blaryjp** [comment]
Awesome! Thank you very much @Rugvip / @henry232323 / @timbonicus !
